File: ios/Sasquatsh/Views/Collections/UserCollectionListView.swift
import SwiftUI

enum CollectionViewMode: String {
    case cards, list
}

struct UserCollectionListView: View {
    let games: [CollectionGame]
    let userName: String
    @State private var searchText = ""
    @State private var viewMode: CollectionViewMode = .cards

    private var filteredGames: [CollectionGame] {
        if searchText.isEmpty { return games }
        return games.filter { $0.gameName.localizedCaseInsensitiveContains(searchText) }
    }

    var body: some View {
        VStack(spacing: 0) {
            // Search + view toggle
            HStack(spacing: 10) {
                HStack {
                    Image(systemName: "magnifyingglass")
                        .foregroundStyle(Color.md3OnSurfaceVariant)
                    TextField("Search \(games.count) games...", text: $searchText)
                    if !searchText.isEmpty {
                        Button { searchText = "" } label: {
                            Image(systemName: "xmark.circle.fill")
                                .foregroundStyle(Color.md3OnSurfaceVariant)
                        }
                    }
                }
                .padding(10)
                .background(Color.md3Surface)
                .clipShape(RoundedRectangle(cornerRadius: MD3Shape.medium))

                // View toggle
                Button {
                    withAnimation(.easeInOut(duration: 0.2)) {
                        viewMode = viewMode == .cards ? .list : .cards
                    }
                } label: {
                    Image(systemName: viewMode == .cards ? "list.bullet" : "square.grid.2x2")
                        .font(.system(size: 18))
                        .frame(width: 40, height: 40)
                        .foregroundStyle(Color.md3OnSecondaryContainer)
                        .background(Color.md3SecondaryContainer)
                        .clipShape(RoundedRectangle(cornerRadius: MD3Shape.small))
                }
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 8)

            if filteredGames.isEmpty {
                Spacer()
                Text(searchText.isEmpty ? "No games" : "No matches")
                    .font(.md3BodyMedium)
                    .foregroundStyle(Color.md3OnSurfaceVariant)
                Spacer()
            } else {
                ScrollView {
                    switch viewMode {
                    case .cards:
                        cardGrid
                    case .list:
                        listView
                    }
                }
            }
        }
        .background(Color.md3SurfaceContainer)
        .navigationTitle("\(userName)'s Games")
        .navigationBarTitleDisplayMode(.inline)
    }

    // MARK: - Card Grid

    private var cardGrid: some View {
        LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible())], spacing: 10) {
            ForEach(filteredGames) { game in
                VStack(spacing: 0) {
                    // Thumbnail
                    if let urlStr = game.thumbnailUrl, let url = URL(string: urlStr) {
                        AsyncImage(url: url) { image in
                            image.resizable().aspectRatio(contentMode: .fill)
                        } placeholder: {
                            Color.md3SurfaceVariant
                        }
                        .frame(height: 100)
                        .clipped()
                    } else {
                        Color.md3SurfaceVariant
                            .frame(height: 100)
                            .overlay {
                                Image(systemName: "dice")
                                    .font(.system(size: 24))
                                    .foregroundStyle(Color.md3OnSurfaceVariant.opacity(0.3))
                            }
                    }

                    VStack(alignment: .leading, spacing: 4) {
                        Text(game.gameName)
                            .font(.md3LabelLarge)
                            .foregroundStyle(Color.md3OnSurface)
                            .lineLimit(2)

                        HStack(spacing: 6) {
                            if let min = game.minPlayers, let max = game.maxPlayers {
                                Label(min == max ? "\(min)p" : "\(min)-\(max)p", systemImage: "person.2")
                                    .font(.system(size: 10))
                                    .foregroundStyle(Color.md3OnSurfaceVariant)
                            }
                            if let time = game.playingTime, time > 0 {
                                Label("\(time)m", systemImage: "clock")
                                    .font(.system(size: 10))
                                    .foregroundStyle(Color.md3OnSurfaceVariant)
                            }
                        }

                        if let rank = game.bggRank, rank > 0 {
                            Text("#\(rank) BGG")
                                .font(.system(size: 10, weight: .medium))
                                .foregroundStyle(Color.md3Tertiary)
                        }
                    }
                    .padding(8)
                    .frame(maxWidth: .infinity, alignment: .leading)
                }
                .background(Color.md3Surface)
                .clipShape(RoundedRectangle(cornerRadius: MD3Shape.medium))
            }
        }
        .padding(.horizontal, 16)
        .padding(.bottom, 16)
    }

    // MARK: - List View

    private var listView: some View {
        LazyVStack(spacing: 4) {
            ForEach(filteredGames) { game in
                HStack(spacing: 12) {
                    if let urlStr = game.thumbnailUrl, let url = URL(string: urlStr) {
                        AsyncImage(url: url) { image in
                            image.resizable().aspectRatio(contentMode: .fill)
                        } placeholder: {
                            Color.md3SurfaceVariant
                        }
                        .frame(width: 40, height: 40)
                        .clipShape(RoundedRectangle(cornerRadius: 4))
                    } else {
                        RoundedRectangle(cornerRadius: 4)
                            .fill(Color.md3SurfaceVariant)
                            .frame(width: 40, height: 40)
                    }

                    VStack(alignment: .leading, spacing: 2) {
                        Text(game.gameName)
                            .font(.md3BodyMedium)
                            .foregroundStyle(Color.md3OnSurface)
                            .lineLimit(1)

                        HStack(spacing: 8) {
                            if let year = game.yearPublished {
                                Text(String(year))
                                    .font(.md3BodySmall)
                            }
                            if let min = game.minPlayers, let max = game.maxPlayers {
                                Text(min == max ? "\(min)p" : "\(min)-\(max)p")
                                    .font(.md3BodySmall)
                            }
                            if let time = game.playingTime, time > 0 {
                                Text("\(time) min")
                                    .font(.md3BodySmall)
                            }
                        }
                        .foregroundStyle(Color.md3OnSurfaceVariant)
                    }

                    Spacer()

                    if let rank = game.bggRank, rank > 0 {
                        Text("#\(rank)")
                            .font(.md3LabelLarge)
                            .foregroundStyle(Color.md3Tertiary)
                    }
                }
                .padding(.horizontal, 16)
                .padding(.vertical, 8)
            }
            .padding(.bottom, 16)
        }
    }
}

File: android/app/src/main/java/com/sasquatsh/app/views/collections/UserCollectionListView.kt
package com.sasquatsh.app.views.collections

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Casino
import androidx.compose.material.icons.filled.Clear
import androidx.compose.material.icons.filled.GridView
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.ViewList
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.sasquatsh.app.models.CollectionGame

private enum class CollectionViewMode { CARDS, LIST }

/**
 * Browse another user's game collection.
 * Supports card grid and list view modes with search filtering.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun UserCollectionListView(
    games: List<CollectionGame>,
    userName: String
) {
    var searchText by remember { mutableStateOf("") }
    var viewMode by remember { mutableStateOf(CollectionViewMode.CARDS) }

    val filteredGames = remember(games, searchText) {
        if (searchText.isEmpty()) games
        else games.filter { it.gameName.contains(searchText, ignoreCase = true) }
    }

    Scaffold(
        topBar = {
            TopAppBar(title = { Text("$userName's Games") })
        }
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .background(MaterialTheme.colorScheme.surfaceContainer)
        ) {
            // Search + view toggle
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp, vertical = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(10.dp)
            ) {
                OutlinedTextField(
                    value = searchText,
                    onValueChange = { searchText = it },
                    placeholder = { Text("Search ${games.size} games...") },
                    leadingIcon = {
                        Icon(Icons.Default.Search, contentDescription = null, tint = MaterialTheme.colorScheme.onSurfaceVariant)
                    },
                    trailingIcon = {
                        if (searchText.isNotEmpty()) {
                            IconButton(onClick = { searchText = "" }) {
                                Icon(Icons.Default.Clear, contentDescription = "Clear", tint = MaterialTheme.colorScheme.onSurfaceVariant)
                            }
                        }
                    },
                    modifier = Modifier.weight(1f),
                    singleLine = true
                )

                IconButton(
                    onClick = {
                        viewMode = if (viewMode == CollectionViewMode.CARDS) CollectionViewMode.LIST
                        else CollectionViewMode.CARDS
                    }
                ) {
                    Icon(
                        if (viewMode == CollectionViewMode.CARDS) Icons.Default.ViewList else Icons.Default.GridView,
                        contentDescription = "Toggle view",
                        tint = MaterialTheme.colorScheme.onSecondaryContainer,
                        modifier = Modifier
                            .size(40.dp)
                            .clip(RoundedCornerShape(8.dp))
                            .background(MaterialTheme.colorScheme.secondaryContainer)
                            .padding(8.dp)
                    )
                }
            }

            if (filteredGames.isEmpty()) {
                Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        if (searchText.isEmpty()) "No games" else "No matches",
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            } else {
                when (viewMode) {
                    CollectionViewMode.CARDS -> CardGrid(filteredGames)
                    CollectionViewMode.LIST -> ListView(filteredGames)
                }
            }
        }
    }
}

// -- Card Grid --

@Composable
private fun CardGrid(games: List<CollectionGame>) {
    LazyVerticalGrid(
        columns = GridCells.Fixed(2),
        contentPadding = PaddingValues(horizontal = 16.dp, vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp)
    ) {
        items(games, key = { it.id }) { game ->
            Column(
                modifier = Modifier
                    .clip(RoundedCornerShape(12.dp))
                    .background(MaterialTheme.colorScheme.surface)
            ) {
                // Thumbnail
                if (game.thumbnailUrl != null) {
                    AsyncImage(
                        model = game.thumbnailUrl,
                        contentDescription = game.gameName,
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(100.dp),
                        contentScale = ContentScale.Crop
                    )
                } else {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(100.dp)
                            .background(MaterialTheme.colorScheme.surfaceVariant),
                        contentAlignment = Alignment.Center
                    ) {
                        Icon(
                            Icons.Default.Casino,
                            contentDescription = null,
                            modifier = Modifier.size(24.dp),
                            tint = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.3f)
                        )
                    }
                }

                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(8.dp)
                ) {
                    Text(
                        game.gameName,
                        style = MaterialTheme.typography.labelLarge,
                        color = MaterialTheme.colorScheme.onSurface,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis
                    )

                    Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                        if (game.minPlayers != null && game.maxPlayers != null) {
                            Row(verticalAlignment = Alignment.CenterVertically) {
                                Icon(
                                    Icons.Default.Person,
                                    contentDescription = null,
                                    modifier = Modifier.size(10.dp),
                                    tint = MaterialTheme.colorScheme.onSurfaceVariant
                                )
                                Spacer(modifier = Modifier.width(2.dp))
                                Text(
                                    if (game.minPlayers == game.maxPlayers) "${game.minPlayers}p"
                                    else "${game.minPlayers}-${game.maxPlayers}p",
                                    style = MaterialTheme.typography.bodySmall.copy(fontSize = 10.sp),
                                    color = MaterialTheme.colorScheme.onSurfaceVariant
                                )
                            }
                        }
                        game.playingTime?.takeIf { it > 0 }?.let { time ->
                            Text(
                                "${time}m",
                                style = MaterialTheme.typography.bodySmall.copy(fontSize = 10.sp),
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }

                    game.bggRank?.takeIf { it > 0 }?.let { rank ->
                        Text(
                            "#$rank BGG",
                            style = MaterialTheme.typography.bodySmall.copy(fontSize = 10.sp),
                            color = MaterialTheme.colorScheme.tertiary
                        )
                    }
                }
            }
        }
    }
}

// -- List View --

@Composable
private fun ListView(games: List<CollectionGame>) {
    LazyColumn(
        contentPadding = PaddingValues(bottom = 16.dp)
    ) {
        items(games, key = { it.id }) { game ->
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp, vertical = 8.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                if (game.thumbnailUrl != null) {
                    AsyncImage(
                        model = game.thumbnailUrl,
                        contentDescription = game.gameName,
                        modifier = Modifier
                            .size(40.dp)
                            .clip(RoundedCornerShape(4.dp)),
                        contentScale = ContentScale.Crop
                    )
                } else {
                    Box(
                        modifier = Modifier
                            .size(40.dp)
                            .clip(RoundedCornerShape(4.dp))
                            .background(MaterialTheme.colorScheme.surfaceVariant)
                    )
                }

                Spacer(modifier = Modifier.width(12.dp))

                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        game.gameName,
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurface,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )

                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        game.yearPublished?.let {
                            Text(
                                "$it",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                        if (game.minPlayers != null && game.maxPlayers != null) {
                            Text(
                                if (game.minPlayers == game.maxPlayers) "${game.minPlayers}p"
                                else "${game.minPlayers}-${game.maxPlayers}p",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                        game.playingTime?.takeIf { it > 0 }?.let { time ->
                            Text(
                                "$time min",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }
                }

                game.bggRank?.takeIf { it > 0 }?.let { rank ->
                    Text(
                        "#$rank",
                        style = MaterialTheme.typography.labelLarge,
                        color = MaterialTheme.colorScheme.tertiary
                    )
                }
            }
        }
    }
}
